### PR TITLE
Api part 2

### PR DESCRIPTION
**Pull request by micahjohnson150**
* Added in some support for rasters. 
* Added in some simplifying of the results from `.all_*`
* Added in `from_unique_entries` which allows users to find options for a query theyre working on . Example:  what dates did specific surface area get collected. 
* Addresses #97

### CODE REVIEW

**Comment by jomey** [comment]
Just a quick reminder to keep track of the changes and maybe start a manual CHANGELOG.md for now until I get back to #85 and automate this?

File: snowexsql/api.py [hunk 1]
```diff
@@ -159,7 +170,7 @@ def all_dates(self):
         with db_session(self.DB_NAME) as (session, engine):
             qry = session.query(self.MODEL.date).distinct()
             result = qry.all()
-        return result
+        return self.retrieve_single_value_result(result)
```
Comment (micah-prime) [inline snowexsql/api.py:173]: Good call

File: snowexsql/api.py [hunk 2]
```diff
@@ -169,6 +180,16 @@ def all_observers(self):
...
+
+    @property
+    def all_units(self):
+        """
+        Return all distinct units in the data
+        """
+        with db_session(self.DB_NAME) as (session, engine):
+            qry = session.query(self.MODEL.units).distinct()
+            result = qry.all()
         return result
```
Comment (micah-prime) [inline snowexsql/api.py:193]:
```suggestion
        return self.retrieve_single_value_result(result)
```

**Review by micah-prime — COMMENTED** [review]
I just had one suggestion that is a result of what I added being out of sync

**Comment by micah-prime** [comment]
@micahjohnson150 are we missing some commits? I don't see what you're referencing with 'Added in some support for rasters.'

**Comment by micahjohnson150** [comment]
@micah-prime you were correct. I hadnt pushed everything apparently and got my edits and commits mixed up lol. This should be everything I have.

File: snowexsql/api.py [hunk 3]
```diff
@@ -270,9 +314,88 @@ class LayerMeasurements(PointMeasurements):
...
+                # Form query and check if the query spans multipl rasters
+                for column in multi_raster_indicators:
+                    values = cls.from_unique_entries([column], **kwargs)
+                    if len(values) > 1:
+                        options = [f"'{v}'" for v in values]
+                        raise TooManyRastersException(f"More than one `{column}` suggests there are multiple raster datasets. "
+                                                      f"Try filter {column} to one of the following values {', '.join(options)}.")
+
+            except Exception as e:
+                session.close()
```
Comment (micah-prime) [inline snowexsql/api.py:357]: I don't think you need to call session lose. I believe the `db_session` should handle that

File: snowexsql/api.py [hunk 4]
```diff
@@ -47,9 +47,8 @@ class BaseDataset:
     # Use this database name
     DB_NAME = DB_NAME
 
-    ALLOWED_QRY_KWARGS = [
-        "site_name", "site_id", "date", "instrument", "observers", "type",
-        "utm_zone", "date_greater_equal", "date_less_equal"
+    ALLOWED_QRY_KWARGS = ["site_name", "site_id", "date", "instrument", "observers", "type",
+        "utm_zone", "date_greater_equal", "date_less_equal", "value_greater_equal", 'value_less_equal',
```
Comment (micah-prime) [inline snowexsql/api.py:51]: Nice addition!

File: requirements.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 utm>=0.5.0,<1.0
 geoalchemy2>=0.6,<1.0
-geopandas>=0.7,<1.0
+geopandas>=0.7,<2.0
```
Comment (micah-prime) [inline requirements.txt:3]: WOOO!

File: snowexsql/api.py [hunk 6]
```diff
@@ -270,9 +314,88 @@ class LayerMeasurements(PointMeasurements):
...
+    @classmethod
+    def from_filter(cls, **kwargs):
+        """
+        Get data for the class by filtering by allowed arguments. The allowed
+        filters are cls.ALLOWED_QRY_KWARGS.
+        """
+        # Grab resolution and remove it from the query kwargs
+        # resolution = kwargs.get("resolution")
+        # if resolution:
+        #     kwargs.pop("resolution")
```
Comment (micah-prime) [inline snowexsql/api.py:370]: Can these comments go?

File: snowexsql/api.py [hunk 7]
```diff
@@ -270,9 +314,88 @@ class LayerMeasurements(PointMeasurements):
...
+                    )
+                )
+                qry = cls.extend_qry(qry, **kwargs)
+                rasters = qry.all()
+
+                # Get the rasterio object of the raster
+                datasets = raster_to_rasterio(rasters)
+
+            except Exception as e:
+                session.close()
```
Comment (micah-prime) [inline snowexsql/api.py:394]: Same comment with this session

**Review by micah-prime — CHANGES_REQUESTED** [review]
Looking good! Couple of minor asks

**Comment by micah-prime** [comment]
> Just a quick reminder to keep track of the changes and maybe start a manual CHANGELOG.md for now until I get back to #85 and automate this?

I'm going to merge #85 for now so we can update that. Then I'll  merge into this branch so we have a changelog to update

**Review by micah-prime — APPROVED** [review]

**Pull request merged**